### PR TITLE
Library for environmental variable access

### PR DESCRIPTION
**Pull request by iwahbe**
Add a library to encapsulate environmental variable declaration, lookup and parsing. 

### Usage
Variables strongly typed and declared in the module scope like so:
```go
// Automatically prefixed with PULUMI_ by default
var Experimental = env.Bool("EXPERIMENTAL", "Enable experimental options and commands")
```

Values are accessed by calling `.Value()` on the module level variable:
```go
if Experimental.Value() { // .Value() returns a bool here
  // do the new exiting thing
} else {
  // do the well tested thing
}
```

Environmental variables can be predicated on other boolean type variables:
```go
var NewThingParam = env.String("NEW_THING", "pass this to the new thing", 
    env.Needs(Expiremental))
```

Predicated variables will only show up as set if there predicate is `true`:

```go
if v := NewThingParam.Value(); v != "" {
  do_thing(v)
} else {
  // do the well tested thing
}
```
The above `if` check is equivalent to 
```go
if v = os.Getenv("PULUMI_NEW_THING"); v != "" && cmduilt.IsTruthy(os.Getenv("PULUMI_EXPERIMENTAL"))
```

This makes marking and unmarking a variable as experimental a 1 line change.

--- 

All declared variables can then be iterated on for documentation with `env.Variables()`. This is how we can use this lib to build documentation for our environmental variables.

#### Immutability 
The library assumes that environmental variables are inherited from the environment. Values are captured during program startup, changes to the environment are not observed. We are not resilient to environmental variables changing out from under us, so this complies with current usage. This assumption can be changed without breaking the API.

#### Known values
The library assumes that the names of environmental variables are known at compile time. If the variable to be looked up is derived from user input, a different mechanism should be used.

### Adoption

Adopting this system over our pulumi/pkg codebase will take a while. My plan is to merge in this PR without moving over our existing env vars from `os.Getenv`. We can then gradually move over existing env vars piecemeal in subsequent PRs. Any new env vars added will use the new system.

Prerequisite for https://github.com/pulumi/pulumi/issues/10747.

### Other Options - Viper
Viper looks like an excellent config retrieval package, but doesn't support at-site documentation. It also has less type safety then I would prefer as a top level api. If it would be helpful, we could switch the underlying mechanism from `os.Lookup` to `viper.Get*` without changing the API this library exposes. I don't think viper is necessary right now, but adopting this library doesn't preclude viper.

### CODE REVIEW

**Comment by RobbieMcKinstry** [comment]
Here's some high-horse commentary, so take it with a grain of disdain:

As a criterion for merging, we should capture in an issue a list of what environment variables are not migrated to this system. IMO having this library merged in without having an actionable next step and a "Definition of Done" leaves room for this to become dead code. In short, my request to "Please open an issue with [this list of environment variables](https://www.pulumi.com/docs/reference/cli/environment-variables/#environment-variables) marked with open checkboxes like so":

```
## Migration Status:

- [ ] PULUMI_AUTOMATION_API_SKIP_VERSION_CHECK
- [ ] PULUMI_ACCESS_TOKEN
...
```

------------

As an aside about Viper (I should have brought this up in ideation when we've spoken about it in the past), I believe in the philosophy that configuration should always be available as both an environment variable and a CLI flag. It's poor ergonomics to have some things as environment variables and others as flags -- both forms of configuration serve a separate purpose; flags are more optimal when working in the terminal while env vars are better for scripts. But there's no reason to inconsistently offer them, forcing users to use flags when they would prefer an environment variable, or environment variables when a flag would suffice.

Viper has this built-in. It respects the "normal order", too:
CLI flags override environment variables.
Environment variables override defaults. (etc)

Of course, this library could offer the same feature if we wanted; I'm not married to Viper, I just like this approach for multiple reasons.

Not going to hold up this PR based on my philosophical leanings, but I'd just posit this as a suggested future long-term route for the CLI.

File: pkg/codegen/report/report.go [hunk 1]
```diff
@@ -29,9 +29,11 @@ import (
 	"github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/syntax"
 	hcl2 "github.com/pulumi/pulumi/pkg/v3/codegen/pcl"
 	"github.com/pulumi/pulumi/pkg/v3/version"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/env"
 )
 
-const ExportTargetDir = "PULUMI_CODEGEN_REPORT_DIR"
+var ExportTargetDir = env.String("CODEGEN_REPORT_DIR",
```
Comment (RobbieMcKinstry) [inline pkg/codegen/report/report.go:35]: For the purpose of hygiene and testability, I'd recommend moving environment variables into an `Environment` struct, and having one global instance of the environment struct rather than many global `Variable` structs. 

Then, it's optional whether the `Environment` object could be initialized in an `init` function or at the top of `main`. 

The primary benefit IMO is that you can inject the `Environment` object into functions that require one or more env vars; this lets you test control flow that uses environment variables without having to test them in your tests (which can be error-prone, since it makes the test impure).

File: sdk/go/common/util/env/env.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+package env
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+)
+
+var Experimental = Bool("EXPERIMENTAL", "Enable experimental options and commands")
```
Comment (RobbieMcKinstry) [inline sdk/go/common/util/env/env.go:27]: Ah! I see that enviroment variables aren't necessarily declared in the library but can occur anywhere in the program.

I would encourage us to have one file that colocates all environment variable declarations. This makes it a lot easier to see what the available configuration options are, and to know immediately where to look when you need a particular env var.

I suppose my suggestion above of a `Environment` struct would necessitate that design.

File: sdk/go/common/util/env/env.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+type Var struct {
```
Comment (RobbieMcKinstry) [inline sdk/go/common/util/env/env.go:29]: Mind adding some comments on public types and methods?

There's a CI lint that I'd like to eventually turn on that forces us to have doc comments on all public identifiers, but we're pretty far away from enabling that lint by default.

File: sdk/go/common/util/env/env.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+	name        string
+	Value       Value
+	Description string
+	options     options
+}
+
+const Prefix = "PULUMI"
```
Comment (RobbieMcKinstry) [inline sdk/go/common/util/env/env.go:36]: Since this appears in the public API, it might be simpler to write this as `PULUMI_` and not have to join it elsewhere. Just a little more explicit for readers of the API.

File: sdk/go/common/util/env/env.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+func (o options) name(underlying string) string {
+	if o.noPrefix {
+		return underlying
+	}
+	return Prefix + "_" + underlying
+}
+
+func Needs(val BoolValue) Option {
```
Comment (RobbieMcKinstry) [inline sdk/go/common/util/env/env.go:65]: Is there no version of this function with a `StringValue` argument?

File: sdk/go/common/util/env/env.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+	Var() Var
+
+	setVar(Var)
+	formattedValue() string
+}
+
+type value struct {
+	underlying string
+	present    bool
+	_var       Var
```
Comment (RobbieMcKinstry) [inline sdk/go/common/util/env/env.go:87]: You can use an anon field here if you want. Having the underscore seems to violate Go style conversions, especially since `_var` is used.

```suggestion
	Var
```

File: sdk/go/common/util/env/env.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+func (v value) String() string {
+	if !v.present {
+		return fmt.Sprintf("%#v: unset", v.Var().Name())
+	}
+	if m := v.missingPrerequs(); m != "" {
+		return fmt.Sprintf("%#v: need %s (%s)", v.Var().Name(), m, v.Var().Value.formattedValue())
+	}
+	return fmt.Sprintf("%#v: %s", v.Var().Name(), v.Var().Value.formattedValue())
+}
```
Comment (RobbieMcKinstry) [inline sdk/go/common/util/env/env.go:98]: Looks like this function can error if the prereqs aren't met. Maybe we should be explicit and return an error here instead?

File: sdk/go/common/util/env/env.go [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+func (v *value) setVar(_var Var) {
+	v._var = _var
+}
```
Comment (RobbieMcKinstry) [inline sdk/go/common/util/env/env.go:102]:
```suggestion
func (v *value) setVar(var Var) {
	v._var = var
}
```

**Review by RobbieMcKinstry — CHANGES_REQUESTED** [review]
The only other high level feedback I have is that introducing an interface type `Value` will lose some type-safety.
The way this is written, the types `BoolValue` and `StringValue` present the type of the underlying value (`bool` or `string`). Once those instances are bound to a local of type `Value` (an interface type), you'll have to ascertain the type of the underlying value using the `type()` method to error-check you've got a flag of the right type. Checking the `type()` method is going to introduce some boilerplate and requires a runtime check. If you eliminate the `Variable` interface entirely, then users are forced to work with `BoolValue` and `StringValue` directly, which requires them to know how they plan to use the flag ahead of time.

If you're using the `interface` type to reduce library boilerplate, then you might be able to use a `Variable` type with a generic argument for `Bool` and `String`.

File: sdk/go/common/util/env/env.go (hunk 5 above)
Comment (iwahbe) [reply sdk/go/common/util/env/env.go:65]: No, for two reasons:
1. I'm not aware of any place where we condition a set of variables on the result of a string value.
2. It's not clear what `Needs(MyString)` would mean. `BoolValue` has a clear truthy coercion. `FOO=1` and `FOO=true` is truthy, `FOO=0` or `unset FOO` is falsey. There isn't a clear surjection from `string` to `bool`. 

If there is a use case where that is necessary, we can add anything way to encode that functionality that puts the onus on the caller to define what true means. The first design I though of was a `When(func() bool)` style api. `Needs(v)` would then be `When(v.Value)`.

File: sdk/go/common/util/env/env.go (hunk 7 above)
Comment (iwahbe) [reply sdk/go/common/util/env/env.go:98]: It doesn't error. It outputs 1 of 3 different strings:
1. `PULUMI_VAR: unset` (if the value was not set)
2. `PULUMI_VAR: needs PULUMI_EXPERIMENTAL (val)` (if the value was set, but isn't used)
3. `PULUMI_VAR: val` (if the value was set and was used)

File: sdk/go/common/util/env/env.go (hunk 8 above)
Comment (iwahbe) [reply sdk/go/common/util/env/env.go]: Unfortunately, this is a parse error. Thats whey I had `_var` instead of `var`. Because the `_` prefix often means "I don't intend to use", I'll change it to `variable`.

File: sdk/go/common/util/env/env.go (hunk 3 above)
Comment (iwahbe) [reply sdk/go/common/util/env/env.go:34]: Done!

File: sdk/go/common/util/env/env.go (hunk 2 above)
Comment (iwahbe) [reply sdk/go/common/util/env/env.go:31]: Yes, environmental variables can be declared anywhere. My assumption is that they will be declared in the module where they are consumed.

I disagree with this. I think that environmental variables should be declared in the module where they are used. We can let the runtime aggregate all declared variables on our behalf. If we want a list of existing env vars, we can call `pulumi env`.

File: pkg/codegen/report/report.go (hunk 1 above)
Comment (iwahbe) [reply pkg/codegen/report/report.go:35]: > For the purpose of hygiene and testability, I'd recommend moving environment variables into an Environment struct, and having one global instance of the environment struct rather than many global Variable structs.

I'm not sure that I see the benefit of a global instance over module level variables here. It also forces each variable to be declared in the same scope, vs where they are used.

> Then, it's optional whether the Environment object could be initialized in an init function or at the top of main.

In general, I think we should avoid tests depending on env vars. If functions need to be explicitly set to up to take the env bundle, they should take just the values they need as arguments. If we want to, we can add support for creating constant values that conform to `StringValue` and `BoolValue`. These can be assigned before testing.

> The primary benefit IMO is that you can inject the Environment object into functions that require one or more env vars; this lets you test control flow that uses environment variables without having to test them in your tests (which can be error-prone, since it makes the test impure).

Can you clarify. I don't think that go supports the kind of injection I'm thinking of here.

**Comment by iwahbe** [comment]
> The only other high level feedback I have is that introducing an interface type `Value` will lose some type-safety. The way this is written, the types `BoolValue` and `StringValue` present the type of the underlying value (`bool` or `string`). Once those instances are bound to a local of type `Value` (an interface type), you'll have to ascertain the type of the underlying value using the `type()` method to error-check you've got a flag of the right type. Checking the `type()` method is going to introduce some boilerplate and requires a runtime check. If you eliminate the `Variable` interface entirely, then users are forced to work with `BoolValue` and `StringValue` directly, which requires them to know how they plan to use the flag ahead of time.

Why would you bind a `BoolValue` to a local of type `Value` if you will need to access the typed value? I introduce the untyped value to allow iterating over declared variables. Since variables are declared where they are used (I consider this a good thing), we need a way to list declared variables for the user. (or query them in any other arbitrary way). This gives a pretty good first approximation of what we want `pulumi env` to do:

```go
for _, v := range env.Variables() {
  fmt.Printf("%s\n", v)
}
```

`env.Variables()` needs to return a type that describes all variables though, and that is `Var`.

> If you're using the `interface` type to reduce library boilerplate, then you might be able to use a `Variable` type with a generic argument for `Bool` and `String`.

`pulumi/{std,pkg}` is compiled with go1.17. We don't have generics. I think we will still need a type to talk about "a value" or "a variable" when aggregating.

**Comment by iwahbe** [comment]
> As a criterion for merging, we should capture in an issue a list of what environment variables are not migrated to this system. IMO having this library merged in without having an actionable next step and a "Definition of Done" leaves room for this to become dead code. In short, my request to "Please open an issue with [this list of environment variables](https://www.pulumi.com/docs/reference/cli/environment-variables/#environment-variables) marked with open checkboxes like so":

The list you link to is slightly different, since it provides variables that are only used in SKDs. It is also out of date. Keeping it up to date is part of the reason why this PR exists. This library is engine focused. We can trivially generated the list of env vars we need to define the follow in the `pkg` and `sdk/go/common` folder.

```sh
rg 'os.(LookupEnv|Getenv)\(.+?\)' -o --no-heading --no-filename | sort | uniq
```

I'll put that into an issue so we can keep track.

File: pkg/codegen/report/report.go (hunk 1 above)
Comment (RobbieMcKinstry) [reply pkg/codegen/report/report.go:35]: Ok, here's my pitch to explain the benefit. It mirrors the definition of the go HTTP server; there's a global instance you can summon when you need it, and ignore when you don't.

Here's the benefit.

Suppose I wanted write a function that checks if pulumi-experimental is enabled and does something.

```golang
func skip_checkpoints() {
   if getModuleLevelEnvVar() == "1"  {
       // ...do the thing
   }
}
```

This requires global state, which makes the test unable to be run in parallel, since it could conflict with other tests also modifying that global state in the same process. 

Instead, you could write:
```golang
func skipCheckpoints(env Environment) {
   if env.Get("PULUMI_EXPERIMENTAL") == "1"  {
       // ...do the thing
   }
}
```
and perform dependency injection of the Environment. Then you'd wrap your original code like this:

```golang
func skipCheckpointsWrapper() {
    var env = getGlobalEnv()
    skipCheckpoints(env)
}
```

The wrapper is what you'd use in your client code, but the version with D.I. is what you'd cover with a unit test. In general, we want unit tests to be pure and total so they can be isolated, so minimizing global state improves testability. That's the general goal of D.I. in this context.

--------

> It also forces each variable to be declared in the same scope, vs where they are used.

Yes. I believe all of our environment variables should be declared in a single location. I don't think we want env variable declarations scattered throughout the code. We should declare all configurables upfront in a single location where possible. This improves discoverability. It's true that means they are "further away" from where they're used, but their location is standardized, and only a single hop ("Jump to Definition") that colocates all of the environment variables. 

> In general, I think we should avoid tests depending on env vars. If functions need to be explicitly set to up to take the env bundle, they should take just the values they need as arguments

I agree that it's true we want to minimize unit tests that exercise features with environment variables. In practice we already have a number of locations that use them. To see those locations, enable this linter in golangci-lint: https://github.com/sivchari/tenv . Those are just the places that set env vars in tests that violate the lints.

As you move up the test pyramid, you'll inevitably have more coarse-grained tests that exercise larger sections of the codebase, some of which will test functionality with configurable toggles. The way to keep this most pure is to inject some `config` struct at each of those sites: i.e. each subsystem of Pulumi is initialized with the set of configurables it needs, following RAII and "fail fast" design methodologies. I agree that if you do that, you won't need to ever set env vars in your tests, because you can always inject your own value. But I worry we aren't far enough along to make that happen, as evidenced by the failing lints.

Does that make sense? I guess I'm trying to offer feedback on two fronts:
1. Why grouping together env vars in a single location is better than spreading them out.
2. Why having a single global Environment instance is better than having a collection of env vars instances (D.I. becomes more straightforward for higher-up-the-pyramid tests).

File: sdk/go/common/util/env/env.go (hunk 2 above)
Comment (RobbieMcKinstry) [reply sdk/go/common/util/env/env.go:31]: IMO I don't believe there's enough benefit to colocating the configurable with the location they're used. It's better to declare them in one location, then perform any validation/munging necessary before initializing the final `Configuration` struct which is known to be well-formed. This allows you to "fail fast" -- return an error to the user before any real computation is done. 

It would be nice if we can put all validation logic for configurables in a single place. That way, deeply nested code isn't trying to check the validity of configuration.

For example, `--parallel=-1` should be checked when the program starts, not when the engine is creating a work queue for the `RegisterResource` requests. Colocating all the logic in place makes that possible (practicing RAII for config).

File: pkg/codegen/report/report.go (hunk 1 above)
Comment (Frassle) [reply pkg/codegen/report/report.go:35]: > Yes. I believe all of our environment variables should be declared in a single location. I don't think we want env variable declarations scattered throughout the code. We should declare all configurables upfront in a single location where possible. This improves discoverability. It's true that means they are "further away" from where they're used, but their location is standardized, and only a single hop ("Jump to Definition") that colocates all of the environment variables.

While I get some of Ian's hesitation towards this, I think this would be more inline with how production ready configuration libraries generally work. 

It is nice being able to define the env vars at the site of use, but 
A) We should probably be adding some friction to adding new envvars, they aren't free and every axis of configuration is an axis to worry about testing and maintaining
B) Being able to inject the environment as a configuration object of all config rather than ad-hoc access to getenv would be a _massive_ unit testing win
C) Up front validation of configuration would be nice. I think there's some debate here about if for example an envvar is set invalid but then doesn't get used wouldn't it be nicer to ignore the error rather than telling the user to fix it, but I'm not sure that's an overall win.

File: pkg/cmd/pulumi/convert_test.go [hunk 9]
```diff
@@ -32,18 +33,20 @@ func TestYamlConvert(t *testing.T) {
...
+	t.Parallel()
+	env := env.NewEnv(env.MapStore{
+		env.Dev.Var().Name(): "true",
+	})
 
 	// Check that we can run convert from PCL to PCL
 	tmp := t.TempDir()
 
-	result := runConvert("pcl_convert_testdata", []string{}, "pcl", "pcl", tmp, true)
+	result := runConvert(env, "pcl_convert_testdata", []string{}, "pcl", "pcl", tmp, true)
```
Comment (iwahbe) [inline pkg/cmd/pulumi/convert_test.go:49]: This is an example of injecting the environment into a function.

An environment is anything that can map a name to a typed value, and it can be easily generated from a `map`.

File: pkg/cmd/pulumi/env.go [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+		Args:  cmdutil.NoArgs,
+		// Since most variables won't be included here, we hide the command. We will
+		// unhide once most existing variables are using the new env var framework and
+		// show up here.
+		Hidden: !env.Experimental.Value(),
+		Run: cmdutil.RunResultFunc(func(cmd *cobra.Command, args []string) result.Result {
+			table := cmdutil.Table{
+				Headers: []string{"Variable", "Description", "Value"},
+			}
+			var foundError bool
```
Comment (iwahbe) [inline pkg/cmd/pulumi/env.go:46]: We don't "fail fast" for invalid environmental variables right now, but we trivially could. I think adding env var "fail fast" should be a separate PR, if we add it, since it would be a breaking change.

File: pkg/codegen/report/report.go (hunk 1 above)
Comment (iwahbe) [reply pkg/codegen/report/report.go:35]: A) I think that this friction is the job of code review. I don't think it needs to be technologically actually difficult. I have centralized env var declaration to a single file.

B) We can do this. I have included an example in a new commit.

C) We can do this. To demonstrate functionality, I have added an experimental `pulumi env` command, which returns a failing exit code when a variable has an invalid value. It warns for suspicious values, like booleans with non-standard false values. 


```
$ PULUMI_DEV=3 pulumi env
warning: PULUMI_DEV: "3" is falsy, but doesn't look like a boolean
Variable                       Description                                         Value
PULUMI_DEV                     Enable features for hacking on pulumi itself        false
PULUMI_EXPERIMENTAL            Enable experimental options and commands            unset
PULUMI_IGNORE_AMBIENT_PLUGINS  Discover additional plugins by examining the $PATH  unset
PULUMI_SKIP_UPDATE_CHECK       Disable checking for a new version of pulumi        unset

File: pkg/cmd/pulumi/convert_test.go (hunk 9 above)
Comment (RobbieMcKinstry) [reply pkg/cmd/pulumi/convert_test.go:49]: Love it!

File: pkg/codegen/report/report.go (hunk 1 above)
Comment (RobbieMcKinstry) [inline pkg/codegen/report/report.go:35]:
```golang
var ExportTargetDir = env.String("CODEGEN_REPORT_DIR",
```
Move this line into `common/env/env.go`?

File: pkg/codegen/report/report.go [hunk 11]
```diff
@@ -205,7 +207,7 @@ func (r *reporter) Close() error {
 func (r *reporter) DefaultExport() error {
 	r.m.Lock()
 	defer r.m.Unlock()
-	dir, ok := os.LookupEnv(ExportTargetDir)
+	dir, ok := ExportTargetDir.Underlying()
```
Comment (RobbieMcKinstry) [inline pkg/codegen/report/report.go:210]: I suppose this should eventually migrate to `env.Global().ExportTargetDir().Underlying()`?
Just wondering

File: sdk/go/common/workspace/plugins.go [hunk 12]
```diff
@@ -1570,8 +1571,7 @@ func getPluginInfoAndPath(
 
 	// If we have a version of the plugin on its $PATH, use it, unless we have opted out of this behavior explicitly.
 	// This supports development scenarios.
-	optOut, isFound := os.LookupEnv("PULUMI_IGNORE_AMBIENT_PLUGINS")
-	includeAmbient := !(isFound && cmdutil.IsTruthy(optOut)) || isBundled
+	includeAmbient := !(env.IgnoreAmbientPlugins.Value()) || isBundled
```
Comment (RobbieMcKinstry) [inline sdk/go/common/workspace/plugins.go:1574]: This is a good example of how the env var library code will be cleaner than what we currently have.

**Review by RobbieMcKinstry — APPROVED** [review]
Really great job!

File: pkg/codegen/report/report.go (hunk 11 above)
Comment (iwahbe) [reply pkg/codegen/report/report.go:210]: `Value.Underlying()` retrieves from the global store by default. I don't see the need to attach it to an explicit `env.Env` unless the `env.Env` is injected (and thus dynamic).

File: pkg/codegen/report/report.go (hunk 1 above)
Comment (iwahbe) [reply pkg/codegen/report/report.go:35]: This environment variable is not actually read by the `pulumi` binary, so I don't want to put it with the rest. This is in the unfortunate part of `pkg/codegen` that is actually a library for downstream consumption.

Because that `codegen/report` is never included, this env doesn't show up in `pulumi env`, which is what we want.

**Comment by iwahbe** [comment]
bors retry

File: pkg/codegen/report/report.go (hunk 11 above)
Comment (RobbieMcKinstry) [reply pkg/codegen/report/report.go:210]: Perfect, sounds good!

File: pkg/codegen/report/report.go (hunk 1 above)
Comment (RobbieMcKinstry) [reply pkg/codegen/report/report.go:35]: Oh, thank makes perfect sense. Brilliant!
